### PR TITLE
Improve exponential backoff

### PR DESCRIPTION
**Pull request by stanlyDoge**
Signed-off-by: Stanislav Knot <sknot@redhat.com>
- Enhancement

### Description
If connection admin client to broker fails, we try simple TCP connection to the broker. If it fails, we force roll the pod because it is in a fail state. Credit for design and code to @tombentley 

### Checklist
- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 1]
```diff
@@ -416,22 +448,64 @@ String podName(Integer podId) {
...
             DescribeClusterResult describeClusterResult = ac.describeCluster();
             KafkaFuture<Node> controller = describeClusterResult.controller();
             controllerNode = controller.get(timeout, unit);
+            restartContext.clearConnectionError();
         } catch (ExecutionException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e.getCause());
+            maybeTcpProbe(podId, e, restartContext);
         } catch (TimeoutException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e);
+            throw new ForceableProblem("lala Error while trying to determine the cluster controller from pod " + podName(podId), e);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:461]: The `lala` probably shouldn't be here ;-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 2]
```diff
@@ -416,22 +448,64 @@ String podName(Integer podId) {
...
 
+    /**
+     * If we've already had trouble connecting to this broker try to probe whether the connection is
+     * open on the broker; if it's not then maybe throw a ForceableProblem to immediately force a restart.
+     * This is an optimization for brokers which don't seem to be running.
+     */
+    private void maybeTcpProbe(int podId, ExecutionException executionException, RestartContext restartContext) throws ForceableProblem {
+        log.info("debuug1 {} {}", restartContext.connectionError(), System.currentTimeMillis());
+        if (restartContext.connectionError() + 120_000 >= System.currentTimeMillis()) {
+            log.info("debuug2");
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:476]: These `debuug` messages are probably also not supposed to be here ;-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 3]
```diff
@@ -193,7 +212,7 @@ public String toString() {
         singleExecutor.schedule(() -> {
             log.debug("Considering restart of pod {} after delay of {} {}", podId, delay, unit);
             try {
-                restartIfNecessary(podId, ctx.backOff.done());
+                restartIfNecessary(podId, ctx.backOff.done(), ctx);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:215]: Do we need to pass both `ctx.backOff.done()` and `ctx`? Cant we pass just `ctx` and get `ctx.backOff.done()` from it later? Or does this have some meaning - e.g. in testing?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 4]
```diff
@@ -416,22 +448,64 @@ String podName(Integer podId) {
...
+    int controller(int podId, Admin ac, long timeout, TimeUnit unit, RestartContext restartContext) throws ForceableProblem, InterruptedException {
         Node controllerNode = null;
         try {
             DescribeClusterResult describeClusterResult = ac.describeCluster();
             KafkaFuture<Node> controller = describeClusterResult.controller();
             controllerNode = controller.get(timeout, unit);
+            restartContext.clearConnectionError();
         } catch (ExecutionException e) {
-            throw new ForceableProblem("Error while trying to determine the cluster controller from pod " + podName(podId), e.getCause());
+            maybeTcpProbe(podId, e, restartContext);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:459]: Why do we use this here and not for Timeout exception?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 5]
```diff
@@ -416,22 +448,64 @@ String podName(Integer podId) {
...
     }
 
+    /**
+     * If we've already had trouble connecting to this broker try to probe whether the connection is
+     * open on the broker; if it's not then maybe throw a ForceableProblem to immediately force a restart.
+     * This is an optimization for brokers which don't seem to be running.
+     */
+    private void maybeTcpProbe(int podId, ExecutionException executionException, RestartContext restartContext) throws ForceableProblem {
+        log.info("debuug1 {} {}", restartContext.connectionError(), System.currentTimeMillis());
+        if (restartContext.connectionError() + 120_000 >= System.currentTimeMillis()) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:475]: Why 120_000 an not operationTimeoutMs?

Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:475]: It is related to AdminClient default timeout (1 minute) + some reasonable offset.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 6]
```diff
@@ -295,13 +315,24 @@ private void closeLoggingAnyError(Admin adminClient) {
...
+
         ForceableProblem(String msg, Throwable cause) {
+            this(msg, cause, false);
+        }
+
+        ForceableProblem(String msg, boolean forceNow) {
+            this(msg, null, forceNow);
+        }
+
+        ForceableProblem(String msg, Throwable cause, boolean immediateForce) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:333]: Can we rename `immediateForce` to `forceNow` to match the field?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 7]
```diff
@@ -416,22 +447,55 @@ String podName(Integer podId) {
...
+     * If we've already had trouble connecting to this broker try to probe whether the connection is
+     * open on the broker; if it's not then maybe throw a ForceableProblem to immediately force a restart.
+     * This is an optimization for brokers which don't seem to be running.
+     */
+    private void maybeTcpProbe(int podId, Exception executionException, RestartContext restartContext) throws ForceableProblem {
+        if (restartContext.connectionError() + numPods * 120_000L >= System.currentTimeMillis()) {
+            try {
+                log.debug("Probing TCP port due to previous problems connecting to pod {}", podId);
+                // do a tcp connect and close (with a short connect timeout)
+                tcpProbe(podName(podId), KafkaCluster.REPLICATION_PORT);
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:475]: If this `tcpProbe()` call does not throw an `IOException` then this method returns normally and we lose the fact that `executionException` was thrown and the caller will say the controller is `-1`. Is that correct, or should we rethrow `executionException` if no `IOException` was thrown?

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests profile=regression testcase=RollingUpdateST

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests profile=regression testcase=io.strimzi.systemtest.RollingUpdateST#testKafkaWontRollUpBecauseTopic,io.strimzi.systemtest.RollingUpdateST#testBrokerConfigurationChangeTriggerRollingUpdate

**Comment by stanlyDoge** [comment]
Failed ST is not related to this PR. It started failing after https://github.com/strimzi/strimzi-kafka-operator/pull/2968

**Pull request merged**